File: 154. Find Minimum in Rotated Sorted Array II.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
    let stk = [nums], ans = [];

    while (stk.length){
        let cur = stk.pop();
        let l=0, r=cur.length-1;

        while (l<r){
            let mid = Math.floor((l+r)/2);

            if (cur[mid]<cur[r]){
                r = mid;
            }else if (cur[mid]>cur[r]){
                l = mid+1;
            }else{
                stk.push(cur.slice(l, mid));
                stk.push(cur.slice(mid+1, r+1))
                break
            }
        }

        if (r===l && cur.length){
            ans.push(cur[l]);
        }
    }
    return Math.min(...ans);
};

File: 154. Find Minimum in Rotated Sorted Array II.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        stk, ans = [nums], []
        
        while stk:
            cur = stk.pop()
            l ,r = 0, len(cur)-1
            
            while l<r:
                mid = (l+r)//2
                
                if cur[mid]<cur[r]:
                    r = mid
                
                elif cur[mid]>cur[r]:
                    l = mid+1
                
                else:
                    stk.append(cur[l:mid])
                    stk.append(cur[mid+1:r+1])
                    break

            if l==r and cur:
                ans.append(cur[l])

        return min(ans)